### ISSUE TITLE
Submodel instruction is copied if submodel is used multiple times

### ISSUE THREAD

**Issue by riazanovskiy** [trevorsandy/lpub3d#291]
New version -- new reports!

### Subject
Consider following file:
```
0 FILE main.ldr
0 Author: LDraw
1 0 20 0 -70 1 0 0 0 1 0 0 0 1 3958.dat
0 STEP
1 16 -30 -48 -120 1 0 0 0 1 0 0 0 1 subModel.ldr
1 16 -30 -48 -20 1 0 0 0 1 0 0 0 1 subModel.ldr
0 STEP
1 1 20 -8 -120 1 0 0 0 1 0 0 0 1 3794a.dat
1 1 20 -16 -70 0 0 1 0 1 0 -1 0 0 3666.dat
1 1 20 -8 -20 1 0 0 0 1 0 0 0 1 3794a.dat
0 STEP
1 16 70 -48 -120 1 0 0 0 1 0 0 0 1 subModel.ldr
1 16 70 -48 -20 1 0 0 0 1 0 0 0 1 subModel.ldr
0 NOFILE 
0 FILE submodel.ldr
0 Author: LDraw
1 4 0 24 0 1 0 0 0 1 0 0 0 1 3062b.dat
0 STEP
1 4 0 0 0 1 0 0 0 1 0 0 0 1 4589.dat
0 NOFILE 
```

Lpub generates 12 page instruction for it, where submodel instruction is copied two times before each usage of that submodel. This was not the case for version 2.3.11
If I add 
```
0 !LPUB CONSOLIDATE_INSTANCE_COUNT GLOBAL FALSE
```
it only affects submodel count (changes x4 to x2) but not the actual pages.

LPub3D 2.3.12 under Windows 10, clean install

**Comment by trevorsandy** [comment]
Thank you for reporting this behaviour. I'll take a look.

Cheers,

**Comment by trevorsandy** [comment]
Corrected. 

The unexpected behaviour was caused by a single `!`. 

In my environment , with Continuous Step Numbers enabled, the behaviour was as expected; and as you described when disabled.

The following shots show the step count for each variation of the Continuous Step Numbers and Consolidate Instance Count configuration pair.

![Screenshot - 24_06_2019 , 08_40_20](https://user-images.githubusercontent.com/13388970/59997455-6b6de580-965d-11e9-83b9-75080d47b940.png)

![Screenshot - 24_06_2019 , 08_39_41](https://user-images.githubusercontent.com/13388970/59997456-6b6de580-965d-11e9-85ab-637ae12a6594.png)

![Screenshot - 24_06_2019 , 08_46_16](https://user-images.githubusercontent.com/13388970/59997457-6b6de580-965d-11e9-8ae5-655f44742868.png)

![Screenshot - 24_06_2019 , 08_46_50](https://user-images.githubusercontent.com/13388970/59997458-6b6de580-965d-11e9-9bcd-25f0ad72762e.png)

Cheers,

**Comment by riazanovskiy** [comment]
Either the CI is broken (I don't think so, the .tar.xz archives are clearly different) or this issue was not fixed :/

I downloaded the latest Arch linux buid from https://github.com/trevorsandy/lpub3d-ci/releases and generated following pdf file

[many-submodels.zip](https://github.com/trevorsandy/lpub3d/files/3320683/many-submodels.zip)

Please look through all pages: they are duplicated. The correct instruction should have 8 pages, not 12

**Comment by riazanovskiy** [comment]
This is clearly a regression: LPub3D 2.3.11 generates a correct 8 page instruction
[many-submodels-11.zip](https://github.com/trevorsandy/lpub3d/files/3320726/many-submodels-11.zip)

**Comment by trevorsandy** [comment]
Actually, it’s the other way around. The behaviour was broken in previous releases - See #264 and #263. My tests and the provided .pdf show the correct number of steps (12) for settings specified in the sample model file.

```
0 !LPUB CONTINUOUS_STEP_NUMBERS GLOBAL FALSE 
0 !LPUB CONSOLIDATE_INSTANCE_COUNT GLOBAL FALSE
```

```
0 FILE main.ldr
0 Author: LDraw
0 !LPUB CONTINUOUS_STEP_NUMBERS GLOBAL FALSE
0 !LPUB CONSOLIDATE_INSTANCE_COUNT GLOBAL FALSE
0 !LPUB RESOLUTION GLOBAL 150 DPI
0 // STEP 1: The base plate
1 0 20 0 -70 1 0 0 0 1 0 0 0 1 3958.dat
0 STEP
0 // STEP 2-6: (5 steps) 2 steps for each occurrence of subModel and the assembled main model
1 16 -30 -48 -120 1 0 0 0 1 0 0 0 1 subModel.ldr
1 16 -30 -48 -20 1 0 0 0 1 0 0 0 1 subModel.ldr
0 STEP
0 // STEP 7: Addition of three bricks
1 1 20 -8 -120 1 0 0 0 1 0 0 0 1 3794a.dat
1 1 20 -16 -70 0 0 1 0 1 0 -1 0 0 3666.dat
1 1 20 -8 -20 1 0 0 0 1 0 0 0 1 3794a.dat
0 STEP
0 // STEP 8-12: (5 steps) 2 steps for each occurrence of subModel and the assembled main model
1 16 70 -48 -120 1 0 0 0 1 0 0 0 1 subModel.ldr
1 16 70 -48 -20 1 0 0 0 1 0 0 0 1 subModel.ldr
0 NOFILE 
0 FILE submodel.ldr
0 Author: LDraw
1 4 0 24 0 1 0 0 0 1 0 0 0 1 3062b.dat
0 STEP
1 4 0 0 0 1 0 0 0 1 0 0 0 1 4589.dat
0 NOFILE 
```

Cheers,

**Comment by trevorsandy** [comment]
I could add another flag to CONSOLIDATE_AFTER_FIRST_OCCURRENCE and set it as the default so as to restore the previous “incorrect” behaviour.

**Comment by riazanovskiy** [comment]
Wow, I didn't think of this. Yes, this is a very nice flag to have.

I see it as following. Continuous step numbering is just a matter of taste. Lego used it in 2018 and didn't use in in 2015.

Not consolidating steps for multiple submodel usages in one step and showing just copies of pages is very strange. This is misleading for instruction user ("Why is it copied and not shown once with "x2"? Do these subassemblies actually differ?"). I have never seen this in an official instruction. 

Consolidation of instructions for same submodel used in different steps can be switched by a flag as it works in 2.3.11. For example, in my instruction I have 4 wheel subassemblies, two of them are added in step 5 and two are added in step 100. So I'd like to have instruction for this submodel wtih "x2" before step 5 and same before step 100. I could achieve this by creating two same mpd subfiles, but as they are literally the same I think it is better to use one subfile in all four places. But I can imagine that someone else would like to tell the instruction user to make four subassemblies, use two of them right now and leave the other two on the table for later.

**Comment by trevorsandy** [comment]
> Not consolidating steps for multiple submodel usages in one step and showing just copies of pages is very strange. This is misleading for instruction user ("Why is it copied and not shown once with "x2"? Do these subassemblies actually differ?"). I have never seen this in an official instruction.

I agree, instance count should not be displayed with option `0 !LPUB CONSOLIDATE_INSTANCE_COUNT GLOBAL FALSE`. I’ll change this behaviour.

> UPD I wrote this before I had seen your edit. With these new commands I guess I'll use

```
CONSOLIDATE_INSTANCE_COUNT_AT_STEP TRUE
CONSOLIDATE_INSTANCE_COUNT_AT_MODEL FALSE
```

That’s correct.

Cheers,

**Comment by elex-at-github** [comment]
The behavior is correct in sub-models, but still incorrect in sub-sub-models.
See the attached example:
pages 4 and 5 should not be there.

![test submodels - all pages](https://user-images.githubusercontent.com/51848093/60395148-68786680-9b2f-11e9-92ee-41051f8066d3.jpg)

**Comment by elex-at-github** [comment]
And to be complete: the .mpd

0 FILE untitled.ldr
0 Untitled
0 Name: Untitled.ldr
0 Author: Alex de Jong
0 Unofficial Model
0 ROTATION CENTER 0 0 0 1 "Custom"
0 ROTATION CONFIG 0 0
0 !LPUB PAGE SIZE GLOBAL 2.9134 4.1339 A7
0 !LPUB CONTINUOUS_STEP_NUMBERS GLOBAL FALSE
0 !LPUB CONSOLIDATE_INSTANCE_COUNT GLOBAL TRUE
1 0 0 0 0 1 0 0 0 1 0 0 0 1 sub1.ldr
1 0 0 -24 0 1 0 0 0 1 0 0 0 1 sub1.ldr
1 0 0 -48 0 1 0 0 0 1 0 0 0 1 sub1.ldr
0 STEP
1 0 60 0 0 1 0 0 0 1 0 0 0 1 sub2.ldr
0
0 NOFILE 
0 FILE sub1.ldr
0 sub1
0 Name: sub1.ldr
0 Author: Alex de Jong
0 Unofficial Model
0 ROTATION CENTER 0 0 0 1 "Custom"
0 ROTATION CONFIG 0 0
1 15 0 0 0 1 0 0 0 1 0 0 0 1 3003.dat
0
0 NOFILE 
0 FILE sub2.ldr
0 sub2
0 Name: sub2.ldr
0 Author: Alex de Jong
0 Unofficial Model
0 ROTATION CENTER 0 0 0 1 "Custom"
0 ROTATION CONFIG 0 0
1 0 0 0 0 1 0 0 0 1 0 0 0 1 sub3.ldr
1 0 0 -24 0 1 0 0 0 1 0 0 0 1 sub3.ldr
1 0 0 -48 0 1 0 0 0 1 0 0 0 1 sub3.ldr
0
0 NOFILE 
0 FILE sub3.ldr
0 sub3
0 Name: sub3.ldr
0 Author: Alex de Jong
0 Unofficial Model
0 ROTATION CENTER 0 0 0 1 "Custom"
0 ROTATION CONFIG 0 0
1 4 0 0 0 1 0 0 0 1 0 0 0 1 3003.dat
0
0
0 NOFILE

**Comment by trevorsandy** [comment]
@elex-at-github, Indeed. 

This behaviour has already been corrected (see #297). I haven't yet pushed it to the production repository.

Here the latest behaviour for the example file you provided.

![Screenshot - 30_06_2019 , 14_22_38](https://user-images.githubusercontent.com/13388970/60396552-4a1c6600-9b43-11e9-92d1-6978a44eb732.png)

![Screenshot - 30_06_2019 , 14_22_58](https://user-images.githubusercontent.com/13388970/60396553-4a1c6600-9b43-11e9-9e68-51fb70ed72fc.png)

![Screenshot - 30_06_2019 , 14_23_16](https://user-images.githubusercontent.com/13388970/60396554-4ab4fc80-9b43-11e9-8957-29c3e2bd1d37.png)

![Screenshot - 30_06_2019 , 14_23_31](https://user-images.githubusercontent.com/13388970/60396555-4ab4fc80-9b43-11e9-853e-c30ae8119589.png)

![Screenshot - 30_06_2019 , 14_23_46](https://user-images.githubusercontent.com/13388970/60396556-4ab4fc80-9b43-11e9-93c8-ec545516f9ba.png)

Cheers,
